src/debug/telemetry/performance_monitor.c

```diff
@@ -397,3 +404,236 @@ int performance_monitor_init(void)
+	size_t entries_cont = self->io_performance_data_bitmap.size;
+
+	for (size_t idx = 0; idx < entries_cont && slot_idx < slots_count; ++idx) {
+		if (perf_bitmap_is_bit_clear(&self->io_performance_data_bitmap, idx))
+			continue;
+		io_global_perf_data->perf_items[slot_idx] = self->io_perf_data[idx];
+		++slot_idx;
+	}
+	io_global_perf_data->perf_item_count = slots_count;
+
+	k_spin_unlock(&self->lock, key);
+	return 0;
+}
+
+static int
+io_perf_monitor_disabled(struct io_perf_monitor_ctx *self)
+{
+	return 0;
+}
+
+static int
+io_perf_monitor_stopped(struct io_perf_monitor_ctx *self)
+{
+	size_t slot_idx = 0;
+	size_t slots_count = self->io_performance_data_bitmap.occupied;
+	size_t entries_cont = self->io_performance_data_bitmap.size;
+
+	for (size_t idx = 0; idx < entries_cont && slot_idx < slots_count; ++idx) {
+		if (perf_bitmap_is_bit_clear(&self->io_performance_data_bitmap, idx))
+			continue;
+
+		self->io_perf_data[idx].data = 0;
+		++slot_idx;
+	}
+
+	return 0;
+}
+
+static int
+io_perf_monitor_started(struct io_perf_monitor_ctx *self)
+{
+	return 0;
+}
+
+static int
+io_perf_monitor_paused(struct io_perf_monitor_ctx *self)
+{
+	return 0;
+}
+
+int io_perf_monitor_set_state(enum ipc4_perf_measurements_state_set state)
+{
+	struct io_perf_monitor_ctx *self = &perf_monitor_ctx;
+	int ret = 0;
+	k_spinlock_key_t key = k_spin_lock(&self->lock);
+
+	switch (state) {
+	case IPC4_PERF_MEASUREMENTS_DISABLED:
+	{
+		ret = io_perf_monitor_disabled(self);
+		break;
+	}
+	case IPC4_PERF_MEASUREMENTS_STOPPED:
+	{
+		ret = io_perf_monitor_stopped(self);
+		break;
+	}
+	case IPC4_PERF_MEASUREMENTS_STARTED:
+	{
+		ret = io_perf_monitor_started(self);
+		break;
+	}
+	case IPC4_PERF_MEASUREMENTS_PAUSED:
+	{
+		ret = io_perf_monitor_paused(self);
+		break;
+	}
+	default:
+		ret = -EINVAL;
+	}
+
+	if (ret == 0)
+		self->state = state;
+
+	k_spin_unlock(&self->lock, key);
+	return ret;
+}
+
+inline enum ipc4_perf_measurements_state_set io_perf_monitor_get_state(void)
+{
+	struct io_perf_monitor_ctx *self = &perf_monitor_ctx;
+
+	return self->state;
+}
+
+int io_perf_monitor_init_data(struct io_perf_data_item **slot_id,
+			      struct io_perf_data_item *init_data)
+{
+	if (!slot_id)
+		return IPC4_ERROR_INVALID_PARAM;
+
+	struct io_perf_monitor_ctx *io_perf_monitor = &perf_monitor_ctx;
+	struct io_perf_data_item *new_slot = io_perf_monitor_get_next_slot(io_perf_monitor);
+
+	if (!new_slot)
+		return IPC4_FAILURE;
+
+	new_slot->id = init_data->id;
+	new_slot->instance = init_data->instance;
+	new_slot->direction = init_data->direction;
+	new_slot->state = init_data->state;
+	new_slot->power_mode = init_data->power_mode;
+	new_slot->is_removed = false;
+	new_slot->data = 0;
+
+	*slot_id = new_slot;
+
+	return 0;
+}
+
+void io_perf_monitor_update_data(struct io_perf_data_item *slot_id, uint32_t increment)
+{
+	if (!slot_id)
+		return;
+
+	struct io_perf_monitor_ctx *self = &perf_monitor_ctx;
+	k_spinlock_key_t key = k_spin_lock(&self->lock);
```

Comment:
I'm not sure about this, this lock was there in reference (not sure what kind of a lock it was), but I think this has potential to disrupt the measured interfaces with a spinlock here. The entries to which we write are just statically defined memory, assigned to a module from performance monitor, and released by the module itself, so I don't think the global lock is needed.

Reply:
CI answered my own question, smokes did not pass. Both this and release slot function cannot be locked. io_perf_monitor_release_slot has locks in the bitmap functions, so it should be safe.